File: app/lib/services/blockchain_service.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:flutter/services.dart';
import 'package:http/http.dart';
import 'package:web3dart/web3dart.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:web3dart/crypto.dart';

class BlockchainService {
  static final BlockchainService _instance = BlockchainService._internal();
  factory BlockchainService() => _instance;
  BlockchainService._internal();

  // Configuration
  static const String _rpcUrl = 'http://192.168.0.215:8545'; // Physical device - direct connection
  // Using computer's IP since ADB port forwarding isn't working
  
  late Web3Client _client;
  late DeployedContract _contract;
  late EthPrivateKey _credentials;
  final _storage = const FlutterSecureStorage();
  
  bool _initialized = false;
  
  Future<void> initialize() async {
    if (_initialized) return;
    
    try {
      print('🔗 Initializing blockchain service...');
      
      // Initialize Web3 client
      _client = Web3Client(_rpcUrl, Client());
      
      // Load contract ABI
      final abiString = await rootBundle.loadString('assets/contracts/BatchTracking.json');
      final abiJson = jsonDecode(abiString);
      
      // Get contract address
      String? contractAddress = await _storage.read(key: 'contract_address');
      
      if (contractAddress == null || contractAddress.isEmpty) {
        throw Exception(
          'Contract address not set. Please deploy contract and set address using setContractAddress()'
        );
      }
      
      print('📄 Contract address: $contractAddress');
      
      _contract = DeployedContract(
        ContractAbi.fromJson(jsonEncode(abiJson['abi']), 'BatchTracking'),
        EthereumAddress.fromHex(contractAddress),
      );
      
      // Load or create credentials
      await _loadOrCreateCredentials();
      
      _initialized = true;
      print('✅ Blockchain service initialized');
      print('👛 Wallet: ${await getWalletAddress()}');
      print('💰 Balance: ${await getBalance()} ETH');
    } catch (e) {
      print('❌ Error initializing blockchain: $e');
      rethrow;
    }
  }
  
  Future<void> _loadOrCreateCredentials() async {
    String? privateKeyHex = await _storage.read(key: 'private_key');
    
    if (privateKeyHex == null) {
      _credentials = EthPrivateKey.createRandom(Random.secure());
      await _storage.write(
        key: 'private_key',
        value: bytesToHex(_credentials.privateKey),
      );
      print('🆕 New wallet created');
    } else {
      _credentials = EthPrivateKey.fromHex(privateKeyHex);
      print('📂 Wallet loaded from storage');
    }
  }
  
  Future<void> setContractAddress(String address) async {
    await _storage.write(key: 'contract_address', value: address);
    _initialized = false;
  }
  
  Future<String> getContractAddress() async {
    return await _storage.read(key: 'contract_address') ?? '';
  }
  
  Future<String> getWalletAddress() async {
    final address = await _credentials.extractAddress();
    return address.hex;
  }
  
  Future<double> getBalance() async {
    try {
      final address = await _credentials.extractAddress();
      final balance = await _client.getBalance(address);
      return balance.getValueInUnit(EtherUnit.ether);
    } catch (e) {
      print('Error getting balance: $e');
      return 0.0;
    }
  }
  
  Future<bool> isConnected() async {
    try {
      await _client.getBlockNumber();
      return true;
    } catch (e) {
      return false;
    }
  }
  
  // ============== BATCH FUNCTIONS ==============
  
  Future<String> createBatch({
    required String batchId,
    required String productName,
    required String variety,
    required int quantity,
    required String unit,
    required String location,
    required DateTime harvestDate,
    required DateTime expiryDate,
    required String nfcTagId,
  }) async {
    await initialize();
    
    try {
      print('📦 Creating batch on blockchain: $batchId');
      
      final function = _contract.function('createBatch');
      
      final transaction = Transaction.callContract(
        contract: _contract,
        function: function,
        parameters: [
          batchId,
          productName,
          variety,
          BigInt.from(quantity),
          unit,
          location,
          BigInt.from(harvestDate.millisecondsSinceEpoch ~/ 1000),
          BigInt.from(expiryDate.millisecondsSinceEpoch ~/ 1000),
          nfcTagId,
        ],
        maxGas: 3000000,
      );
      
      final txHash = await _client.sendTransaction(
        _credentials,
        transaction,
        chainId: 1337,
      );
      
      print('📤 Transaction sent: $txHash');
      await _waitForTransaction(txHash);
      print('✅ Batch created successfully');
      
      return txHash;
    } catch (e) {
      print('❌ Error creating batch: $e');
      rethrow;
    }
  }
  
  Future<Map<String, dynamic>?> getBatch(String batchId) async {
    await initialize();
    
    try {
      final function = _contract.function('getBatch');
      
      final result = await _client.call(
        contract: _contract,
        function: function,
        params: [batchId],
      );
      
      if (result.isEmpty) return null;
      
      return {
        'productName': result[0] as String,
        'variety': result[1] as String,
        'quantity': (result[2] as BigInt).toInt(),
        'unit': result[3] as String,
        'location': result[4] as String,
        'harvestDate': DateTime.fromMillisecondsSinceEpoch(
          (result[5] as BigInt).toInt() * 1000,
        ),
        'expiryDate': DateTime.fromMillisecondsSinceEpoch(
          (result[6] as BigInt).toInt() * 1000,
        ),
        'nfcTagId': result[7] as String,
        'creator': (result[8] as EthereumAddress).hex,
        'timestamp': DateTime.fromMillisecondsSinceEpoch(
          (result[9] as BigInt).toInt() * 1000,
        ),
      };
    } catch (e) {
      print('❌ Error getting batch: $e');
      return null;
    }
  }
  
  Future<String> addQualityMetric({
    required String metricId,
    required String batchId,
    required String metricType,
    required String value,
    required String unit,
  }) async {
    await initialize();
    
    try {
      final function = _contract.function('addQualityMetric');
      
      final transaction = Transaction.callContract(
        contract: _contract,
        function: function,
        parameters: [metricId, batchId, metricType, value, unit],
        maxGas: 2000000,
      );
      
      final txHash = await _client.sendTransaction(
        _credentials,
        transaction,
        chainId: 1337,
      );
      
      await _waitForTransaction(txHash);
      return txHash;
    } catch (e) {
      print('❌ Error adding quality metric: $e');
      rethrow;
    }
  }
  
  Future<Map<String, dynamic>> authenticateNFC({
    required String authId,
    required String batchId,
    required String nfcTagId,
    required String location,
  }) async {
    await initialize();
    
    try {
      print('🔐 Authenticating NFC on blockchain...');
      
      final function = _contract.function('authenticateNFC');
      
      final transaction = Transaction.callContract(
        contract: _contract,
        function: function,
        parameters: [authId, batchId, nfcTagId, location],
        maxGas: 2000000,
      );
      
      final txHash = await _client.sendTransaction(
        _credentials,
        transaction,
        chainId: 1337,
      );
      
      final receipt = await _waitForTransaction(txHash);
      
      // Check authentication result
      final isValid = await _checkAuthResult(batchId, nfcTagId);
      
      print('✅ NFC authenticated. Valid: $isValid');
      
      return {
        'txHash': txHash,
        'isValid': isValid,
        'blockNumber': receipt.blockNumber.blockNum.toInt(),
      };
    } catch (e) {
      print('❌ Error authenticating NFC: $e');
      rethrow;
    }
  }
  
  Future<bool> _checkAuthResult(String batchId, String nfcTagId) async {
    try {
      final batch = await getBatch(batchId);
      return batch?['nfcTagId'] == nfcTagId;
    } catch (e) {
      return false;
    }
  }
  
  Future<int> getTotalBatches() async {
    await initialize();
    
    try {
      final function = _contract.function('getTotalBatches');
      final result = await _client.call(
        contract: _contract,
        function: function,
        params: [],
      );
      return (result[0] as BigInt).toInt();
    } catch (e) {
      return 0;
    }
  }
  
  Future<List<String>> getAllBatchIds() async {
    await initialize();
    
    try {
      final function = _contract.function('getAllBatchIds');
      final result = await _client.call(
        contract: _contract,
        function: function,
        params: [],
      );
      return (result[0] as List).map((e) => e.toString()).toList();
    } catch (e) {
      return [];
    }
  }
  
  Future<TransactionReceipt> _waitForTransaction(String txHash) async {
    const maxAttempts = 60;
    const delay = Duration(seconds: 2);
    
    print('⏳ Waiting for confirmation...');
    
    for (var i = 0; i < maxAttempts; i++) {
      try {
        final receipt = await _client.getTransactionReceipt(txHash);
        if (receipt != null) {
          print('✓ Confirmed in block ${receipt.blockNumber}');
          return receipt;
        }
      } catch (e) {
        // Continue waiting
      }
      await Future.delayed(delay);
    }
    
    throw Exception('Transaction timeout');
  }
  
  void dispose() {
    _client.dispose();
  }
}

File: src/services/blockchainService.ts
import { ethers } from 'ethers';
import BatchTrackingABI from '../contracts/BatchTracking.json';

interface BlockchainConfig {
  rpcUrl: string;
  contractAddress: string;
  chainId: number;
}

class BlockchainService {
  private static instance: BlockchainService;
  private provider: ethers.JsonRpcProvider | null = null;
  private signer: ethers.Wallet | ethers.HDNodeWallet | null = null;
  private contract: ethers.Contract | null = null;
  private config: BlockchainConfig = {
    rpcUrl: 'http://127.0.0.1:8545', // Local Ganache
    contractAddress: '', // Will be set after deployment
    chainId: 1337, // Ganache chain ID
  };

  private constructor() {}

  static getInstance(): BlockchainService {
    if (!BlockchainService.instance) {
      BlockchainService.instance = new BlockchainService();
    }
    return BlockchainService.instance;
  }

  async initialize(contractAddress: string, privateKey?: string): Promise<void> {
    try {
      console.log('🔗 Initializing blockchain service...');
      
      this.config.contractAddress = contractAddress;
      
      // Initialize provider
      this.provider = new ethers.JsonRpcProvider(this.config.rpcUrl);
      
      // Initialize signer
      if (privateKey) {
        this.signer = new ethers.Wallet(privateKey, this.provider);
      } else {
        // Generate random wallet for testing
        const wallet = ethers.Wallet.createRandom();
        this.signer = wallet.connect(this.provider);
        console.log('🆕 New wallet created:', this.signer.address);
        console.log('⚠️  Save this private key:', wallet.privateKey);
      }
      
      // Initialize contract
      this.contract = new ethers.Contract(
        this.config.contractAddress,
        BatchTrackingABI.abi,
        this.signer
      );
      
      console.log('✅ Blockchain service initialized');
      console.log('📄 Contract:', this.config.contractAddress);
      console.log('👛 Wallet:', this.signer.address);
      
      const balance = await this.getBalance();
      console.log('💰 Balance:', balance, 'ETH');
    } catch (error) {
      console.error('❌ Error initializing blockchain:', error);
      throw error;
    }
  }

  async getBalance(): Promise<string> {
    if (!this.provider || !this.signer) throw new Error('Not initialized');
    const balance = await this.provider.getBalance(this.signer.address);
    return ethers.formatEther(balance);
  }

  getWalletAddress(): string {
    if (!this.signer) throw new Error('Not initialized');
    return this.signer.address;
  }

  async isConnected(): Promise<boolean> {
    try {
      if (!this.provider) return false;
      await this.provider.getBlockNumber();
      return true;
    } catch {
      return false;
    }
  }

  // ============== BATCH FUNCTIONS ==============

  async createBatch(
    batchId: string,
    productName: string,
    variety: string,
    quantity: number,
    unit: string,
    location: string,
    harvestDate: Date,
    expiryDate: Date,
    nfcTagId: string
  ): Promise<string> {
    if (!this.contract) throw new Error('Contract not initialized');

    try {
      console.log('📦 Creating batch on blockchain:', batchId);

      const tx = await this.contract.createBatch(
        batchId,
        productName,
        variety,
        BigInt(quantity),
        unit,
        location,
        BigInt(Math.floor(harvestDate.getTime() / 1000)),
        BigInt(Math.floor(expiryDate.getTime() / 1000)),
        nfcTagId,
        {
          gasLimit: 3000000,
        }
      );

      console.log('📤 Transaction sent:', tx.hash);
      const receipt = await tx.wait();
      console.log('✅ Batch created in block:', receipt.blockNumber);

      return tx.hash;
    } catch (error) {
      console.error('❌ Error creating batch:', error);
      throw error;
    }
  }

  async getBatch(batchId: string): Promise<any> {
    if (!this.contract) throw new Error('Contract not initialized');

    try {
      const result = await this.contract.getBatch(batchId);

      return {
        productName: result[0],
        variety: result[1],
        quantity: Number(result[2]),
        unit: result[3],
        location: result[4],
        harvestDate: new Date(Number(result[5]) * 1000),
        expiryDate: new Date(Number(result[6]) * 1000),
        nfcTagId: result[7],
        creator: result[8],
        timestamp: new Date(Number(result[9]) * 1000),
      };
    } catch (error) {
      console.error('❌ Error getting batch:', error);
      return null;
    }
  }

  async addQualityMetric(
    metricId: string,
    batchId: string,
    metricType: string,
    value: string,
    unit: string
  ): Promise<string> {
    if (!this.contract) throw new Error('Contract not initialized');

    try {
      const tx = await this.contract.addQualityMetric(
        metricId,
        batchId,
        metricType,
        value,
        unit,
        { gasLimit: 2000000 }
      );

      console.log('📊 Quality metric added:', tx.hash);
      await tx.wait();

      return tx.hash;
    } catch (error) {
      console.error('❌ Error adding quality metric:', error);
      throw error;
    }
  }

  async authenticateNFC(
    authId: string,
    batchId: string,
    nfcTagId: string,
    location: string
  ): Promise<{ txHash: string; isValid: boolean }> {
    if (!this.contract) throw new Error('Contract not initialized');

    try {
      console.log('🔐 Authenticating NFC on blockchain...');

      const tx = await this.contract.authenticateNFC(
        authId,
        batchId,
        nfcTagId,
        location,
        { gasLimit: 2000000 }
      );

      console.log('📤 Transaction sent:', tx.hash);
      const receipt = await tx.wait();

      // Check if authentication was successful
      const batch = await this.getBatch(batchId);
      const isValid = batch && batch.nfcTagId === nfcTagId;

      console.log('✅ NFC authenticated. Valid:', isValid);

      return {
        txHash: tx.hash,
        isValid,
      };
    } catch (error) {
      console.error('❌ Error authenticating NFC:', error);
      throw error;
    }
  }

  async getTotalBatches(): Promise<number> {
    if (!this.contract) throw new Error('Contract not initialized');

    try {
      const total = await this.contract.getTotalBatches();
      return Number(total);
    } catch (error) {
      console.error('❌ Error getting total batches:', error);
      return 0;
    }
  }

  async getAllBatchIds(): Promise<string[]> {
    if (!this.contract) throw new Error('Contract not initialized');

    try {
      const ids = await this.contract.getAllBatchIds();
      return ids;
    } catch (error) {
      console.error('❌ Error getting batch IDs:', error);
      return [];
    }
  }

  async getBatchQualityMetrics(batchId: string): Promise<string[]> {
    if (!this.contract) throw new Error('Contract not initialized');

    try {
      const metrics = await this.contract.getBatchQualityMetrics(batchId);
      return metrics;
    } catch (error) {
      console.error('❌ Error getting quality metrics:', error);
      return [];
    }
  }

  async getBatchAuthentications(batchId: string): Promise<string[]> {
    if (!this.contract) throw new Error('Contract not initialized');

    try {
      const auths = await this.contract.getBatchAuthentications(batchId);
      return auths;
    } catch (error) {
      console.error('❌ Error getting authentications:', error);
      return [];
    }
  }

  // ============== EVENT LISTENERS ==============

  onBatchCreated(callback: (batchId: string, productName: string, creator: string) => void): void {
    if (!this.contract) throw new Error('Contract not initialized');

    this.contract.on('BatchCreated', (batchId, productName, creator) => {
      console.log('📦 Batch created event:', { batchId, productName, creator });
      callback(batchId, productName, creator);
    });
  }

  onNFCAuthenticated(
    callback: (authId: string, batchId: string, nfcTagId: string, isValid: boolean) => void
  ): void {
    if (!this.contract) throw new Error('Contract not initialized');

    this.contract.on('NFCAuthenticated', (authId, batchId, nfcTagId, isValid) => {
      console.log('🔐 NFC authenticated event:', { authId, batchId, nfcTagId, isValid });
      callback(authId, batchId, nfcTagId, isValid);
    });
  }

  removeAllListeners(): void {
    if (this.contract) {
      this.contract.removeAllListeners();
    }
  }
}

export default BlockchainService;
